ARC-0012: Standard for Claimable Algorand Standard Assets (ASA) (#130)
algorandfoundation/ARCs: ARCs/arc-0012.md

## Reference Implementation

### Source code

[https://github.com/joe-p/Claimable-ASA-PoC](https://github.com/joe-p/Claimable-ASA-PoC)

### Live demo on Testnet

[https://claimable-asa-demo.netlify.app/](https://claimable-asa-demo.netlify.app/).

### Overview

To demonstate how to interact with a claimable ASAs account, a simple application has been created with MyAlgo integration and the following capabilities:

1. Displaying a connected wallet account's current assets
2. Deriving the connected account's *corresponding claimable ASAs account* using the standard TEAL smart signature code
3. Displaying any ASAs in the claimable ASAs account with the ability to claim them into the connected wallet account
4. A smart sending tool to transfer ASAs from the connected wallet account to an intended receiver, dynamically diverting the transfer to the receiver's corresponding claimable ASAs account if the receiver is not opted in to the ASA.*

With this toolset it is possible to view one's own claimable ASAs account, claim ASAs from it, and automatically utilize other people's claimable ASAs account to send them any ASA.

### Burning Tokens

> Additionally, the smart sending tool can be used to burn ASAs by entering the global Zero Address `AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ` as the intended recipient.  The app will automatically re-route the ASA to the unique "UN-claimable" ASAs account corresponding to the Zero Address.  An ASA sent here can never be claimed by anyone, rendering the transferred quantity permanently irretrievable and effectively "burned" in a verifiable way.

Reviewer: So there will 1 Burner Account for 1 ASA?

Reasoning on "ASA irreversible burn" I was inclined to have 1 single "ASA skin", well known in the ecosystem, that could only opt-in into ASA with no Clawback Address, so that anyone would now that is ASA is sent to the known App ID `123` is burned for sure (without having to check the specific burner logic).

WDYT?

Author: There will be one smart signature that can be used to burn any ASA. To verify how much of an ASA is burned, one would need to simply check the balance of that one account. The logic for this contract will show it is claimable by the zero address `AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAY5HFKQ`, but since this account cannot send transactions it is considered burnt.

Reviewer: Can the bruner address opt-in ASAs with non-null Clawback Address?

Reviewer 2: No, but this intent of this ARC is not to neutralize Clawback as a protocol-level feature.

Reviewer: I'm not proposing to neutralize it. I'm just trying to understand if the proposed solution guarantees an _irreversible_ burning of ASAs.

Reviewer 2: We can add some language to clarify explicitly that it is not possible to permanently burn a clawback-enabled ASA under any circumstances (as far as I am aware). This should help avoid any misunderstanding about the fact that this logicSig mechanism cannot override clawback.

authors: Brian Whippo (@silentrhetoric), Joe Polny (@joe-p)
discussions-to: N/A
status: Draft
type: Standards Track
category: ARC
created: 2022-09-05
---

# Claimable Algorand Standard Assets (ASAs)

## Summary

A standard for TEAL contract accounts to serve as a pending claimable ASAs account for Algorand Standard Assets for an intended receiver that is not opted in to the ASA.

## Abstract

The goal of this standard is to establish a standard in the Algorand ecosytem by which ASAs can be sent to an intended receiver even if their account is not opted in to the ASA.  

TEAL stateless smart signature contract accounts corresponding to any Algorand account serve as a claimable ASAs account with logic that permits opting into incoming ASAs and subsequently for only the intended receiver to claim the ASA into their normal account.  A single associated stateful application on the network enables a cost--the minimum balance requirement to create an account and hold an ASA--to be refunded to the ASA creator when the asset is claimed by the receiver.

> If integrated into ecosystem technologies including wallets, epxlorers, and dApps, this standard can provide enhanced capabilities around ASAs which are otherwise strictly bound at the protocol level to require opting in to be received.  This also enables the ability to "burn" ASAs by sending them to the claimable ASAs account associated with the global Zero Address.

Reviewer 3: ```suggestion
If integrated into ecosystem technologies including wallets, explorers, and dApps, this standard can provide enhanced capabilities around ASAs which are otherwise strictly bound at the protocol level to require opting in to be received.  This also enables the ability to "burn" ASAs by sending them to the claimable ASAs account associated with the Zero Address.
```

## Motivation

> Algorand requires users to opt in to receive any ASA, a fact which simultaneously:

Reviewer 3: ```suggestion
Algorand requires accounts to opt in to receive any ASA, a fact which simultaneously:
```

> 1. Grants account holders fine-grained control over the contents of their account by allowing them to select what assets to allow and preventing any unwanted tokens from appearing in their balance.

Reviewer 3: ```suggestion
1. Grants account holders fine-grained control over their holdings by allowing them to select which assets to allow and preventing receipt of unwanted tokens.
```

> 2. Annoys users when transactions fail and must be resubmitted after completing an opt in transaction--or asking someone else to do so.  This routine is jarring to people who are new to Algorand or are accutomed other blockchains without this requirement.

Reviewer 3: ```suggestion
2. Frustrates users and developers when accounting for this requirement especially since other blockchains do not have this requirement.
```

- **Claimable ASAs Account**: An Algorand contract account derived deterministically from the account of an intended ASA receiver plus a stateless TEAL smart signature that approves transactions originating from it.  
- **dApp**: A decentralized application frontend, interpreted here to mean an off-chain frontend (a webapp, native app, etc.) that interacts with applications on the blockchain.
- **Explorer**: An off-chain application that allows browsing the blockchain, showing details of transactions.
- **Wallet**: An off-chain application that stores secret keys for on-chain accounts and can display and sign transactions for these accounts.
- **Mainnet ID**: The ID for the application that should be called upon claiming an asset on mainnet
- **Testnet ID**: The ID for the application that should be called upoin claiming an asset on testnet
- **Minimum Balance Requirement (MBR)**: The minimum amount of Algos which must be held by an account on the ledger, which is currently 0.1A + 0.1A per ASA opted into.

### TEAL Smart Contracts

> There are two smart contracts being used for ARC-0012. A [smart signature](../assets/arc-0012/claimable_lsig.teal) and a [stateful application](../assets/arc-0012/claim_app.teal). In summary, the smart signature is the account used for holding the ASA and handling opt-in and claim logic. The stateful application is used atomically upon claim to ensure the creator of the ASA is paid 0.2 ALGO.

Reviewer 3: ```suggestion
There are two smart contracts specified for ARC-0012:
1. A [smart signature](../assets/arc-0012/claimable_lsig.teal) is the account used for holding the ASA and handling opt-in and claim logic. 
2. A [stateful application](../assets/arc-0012/claim_app.teal).  is used atomically upon claim to ensure the creator of the ASA is paid 0.2 ALGO.
```

##### handle_claim_payment

Logic for handling the payment of 0.2 ALGO that is sent back to the ASA creator upon claiming of the ASA by the intended receiver. This logic relies on an atomic app call to the stateful application to verify the receiver of the payment is the asset creator. A stateful application must be used here because stateless smart signatures cannot get the creator address of an ASA.

##### handle_optin

Logic for verifying an opt-in transaction for the claimable ASAs account. The receiver must be the sender (which is the smart signature address) and the transaction MUST NOT be closing out the asset.

##### handle_claim_axfer

Logic for transfering the claimable ASA to the intended receivers account. This transaction must be grouped with a transaction handled by `handle_claim_payment` and the stateful application.

##### handle_axfer

Logic for determining whether the logic should branch to `handle_claim_axfer` or `handle_optin`.

##### main

Logic that contains necessary assertions, such as verifying the account is not being closed or rekey. Ultimately routes to `handle_claim_payment`, `handle_axfer` or `err`.

#### Smart Contract (Stateful App)

> The application is used to ensure the creator is payed upon ASA claiming. A stateful application must be used to get the creator of the ASA. The application logic is broken down into multiple sections.

Reviewer 3: ```suggestion
The application is used to ensure the creator is refunded upon ASA claiming. A stateful application must be used to get the creator of the ASA. The application logic is broken down into multiple sections.
```
